charmcraft/config.py

```diff
@@ -18,167 +18,195 @@
+        """
+        try:
+            return self.__root__[part_name]
+        except KeyError:
+            return None
+
+
+class CharmhubConfig(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Represents top-level charmhub object."""
+
+    api_url: pydantic.StrictStr = "https://api.charmhub.io"
+    storage_url: pydantic.StrictStr = "https://storage.snapcraftcontent.com"
+
+    @pydantic.validator("api_url", "storage_url")
+    def validate_urls(cls, url):
+        """Verify valid URLs are used for api and storage."""
+        return check_url(url)
+
+
+class Project(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
     """Configuration for all project-related options, used internally."""
 
-    dirpath = attr.ib(default=None)
-    config_provided = attr.ib(default=None)
-    started_at = attr.ib(default=None)
-
-
-@attr.s(kw_only=True, frozen=True)
-class Config:
-    """Root of all the configuration."""
-
-    charmhub = attr.ib(default={}, converter=CharmhubConfig.from_dict)
-    parts = attr.ib(default={}, converter=BasicPrime.from_dict)
-    type = attr.ib(default=None)
-
-    # this item is provided by the code itself, not the user, as convenience for the
-    # rest of the code
-    project = attr.ib(default=None)
-
-
-CONFIG_SCHEMA = {
-    "type": "object",
-    "properties": {
-        "type": {"type": "string", "enum": ["charm", "bundle"]},
-        "charmhub": {
-            "type": "object",
-            "properties": {
-                "api_url": {"type": "string", "format": "url"},
-                "storage_url": {"type": "string", "format": "url"},
-            },
-            "additionalProperties": False,
-        },
-        "parts": {
-            "type": "object",
-            "properties": {
-                "bundle": {
-                    "type": "object",
-                    "properties": {
-                        "prime": {
-                            "type": "array",
-                            "items": {
-                                "type": "string",
-                                "format": "relative_path",
-                            },
-                        },
-                    },
-                },
-            },
-        },
-    },
-    "required": ["type"],
-    "additionalProperties": False,
-}
+    dirpath: pydantic.DirectoryPath
+    content: Dict[str, Any] = {}
+    config_provided: bool = False
+    started_at: datetime.datetime
+
+    @pydantic.validator("dirpath")
+    def validate_path(cls, path):
+        """Verify valid URLs are used for api and storage."""
+        if not isinstance(path, pathlib.Path):
+            raise ValueError("invalid path")
+        return path
+
+
+class Config(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Charmcraft config.
+
+    :ivar type: If none, no charmcraft.yaml is present.
```

Comment:
It could be worth to list the valid types.

Reply:
Simplified the documentation here so we can call schema().